### ISSUE TITLE
How to handle empty selections in the menu?

### ISSUE THREAD

**Issue by heatherleaf** [MUSTE-Project/MULLE#43]
Empty selections `{}` cannot be shown in the UI:
```
{}: flickan slår på en dator
flickor slår på en dator
flickorna slår på en dator
```

The reason why the selection becomes empty is because the node sequence is the same:
```
{}: flickan slår på en dator     girl_N hit_V+switch_on_V on_Prep+switch_on_V aSg_Det computer_N
flickor slår på en dator         girl_N hit_V+switch_on_V on_Prep+switch_on_V aSg_Det computer_N
flickorna slår på en dator       girl_N hit_V+switch_on_V on_Prep+switch_on_V aSg_Det computer_N
```

In this case we would like the menu to be in selection `{0-1}`, because the word "flickan" is changed.

**Comment by fredefox** [comment]
I'll investigate if we can truly get rid of the annotated linearization. In that case I would say that from the point of view of complexity it would be nice to do. I used type-families and type-classes to make the two implementations share functionality. If we just have one data-type we can get rid of all this. The `IsToken` class is similar in style to @daherb's tree class (`TreeC`). That class was nice because we we're exposing two tree representations. Now `LTree` is just an implementation detail. That is to say, its not exported.

**Comment by heatherleaf** [comment]
Insertion is a potential problem if we don't have annotated linearisations. 

### Example 1

Currently, we click before "flickan" to insert "en":

    {0-0}  [] flickan slår  -->  [en] flicka slår

With the "new" suggestion (no annotations), we get the following:

    {0-1}  [flickan] slår  -->  [en flicka] slår

The problem is that the latter formulation doesn't give any information about where to click for insertion - i.e., without annotations we cannot see the difference between these possible versions:

1. `{0-1} [flickan] slår  --> [en flicka] slår`
2. `{0-0,0-1} [[] flickan] slår  --> [[en] flicka] slår`
3. `{0-1,1-1} [flickan []] slår  --> [en [flicka]] slår`
4. `{0-0,0-1,1-1} [[] flickan []] slår  --> [[en] [flicka]] slår`

It's version (2) we want, i.e. we should be able to click before and on "flickan", but not after.

### Example 2

Let's assume that we have a language where nouns are inflected depending on if there is an adjective or not (I'm sure there are such languages). Also, some adjectives are pre-noun, others are post-noun (similar to e.g. French). Here's how it could look like:

With annotated linearisations:

    {1-1}  en [] flicka slår  -->  en [stor] Xflicka slår
    {2-2}  en flicka [] slår  -->  en flickaX [glad] slår

Without annotations:

    {1-2}  en [flicka] slår  -->  en [stor Xflicka] slår
    {1-2}  en [flicka] slår  -->  en [flickaX glad] slår

What I would like to see:

    {1-1,1-2}  en [[] flicka] slår  -->  en [[stor] Xflicka] slår
    {1-2,2-2}  en [flicka []] slår  -->  en [flickaX [glad]] slår

**Comment by heatherleaf** [comment]
This means that we have to call `alignSequences` twice, and then combine the two edit sequences.  Somewhere here:

https://github.com/MUSTE-Project/MULLE/blob/8c20a585f2a9d9764bc9260240dd04cf8318263a/muste/src/Muste/Menu/Internal.hs#L209-L211

Here's an idea:

```
let nodeedits = alignSequences oldnodes newnodes
let wordedits = alignSequences oldwords newwords
let edits = nodeedits <+> wordedits
```

`alignSequences` returns a list of `Edit a`. This should anyway really be a `Set (Edit a)`, because we don't want duplicates. But the types differ:

```
nodeedits :: Set (Edit Node)
wordedits :: Set (Edit Word)
```

However, the type `a` is just intermediary, because `splitAlignments` remove all `a` anyway. So, here's a suggestion (not tested):

```
collectMenuItems ctxt substs = ... $ do
    ...
    let nodeedits :: Set Edit = alignSequences oldnodes newnodes
    let wordedits :: Set Edit = alignSequences oldwords newwords
    let edits :: Set Edit = nodeedits <+> wordedits
    let (oldselection, newselection) = splitAlignments edits
    ...

type Edit a = (([a], Int, Int), ([a], Int, Int))
type Edit = ((Int, Int), (Int, Int))

alignSequences :: Eq a => [a] -> [a] -> Set Edit
alignSequences xs ys = Set.fromList $ cleanEdits $ mergeEdits $ snd $ a Array.! (0,0)
   where ...

cleanEdits :: [Edit a] -> [Edit]
cleanEdits eds = [ ((i,j),(k,l)) | ((_,i,j),(_,k,l)) <- eds ]
```

(I know that I'm illegal here, with `Edit` and `Edit a`)

Note that `mergeEdits` still needs the information about `a`, but after that we can remove that info.

I don't know if this will work, and how well. So I guess best is to put it in a separate branch for testing.

**Comment by heatherleaf** [comment]
Hopefully solved in f68b62bdd21c4cd7a1c54ae5f7b66233b141de0c
